apps/framework-cli/src/infrastructure/redis/connection.rs

```diff
@@ -0,0 +1,140 @@
+    /// The system is currently attempting to reestablish a lost connection.
+    Reconnecting,
+}
+
+/// A thread-safe wrapper around Redis ConnectionManager.
+///
+/// This struct provides a wrapper around the Redis ConnectionManager that
+/// ensures thread safety and adds features like connection health
+/// monitoring and automatic reconnection with exponential backoff.
+#[derive(Clone)]
+pub struct ConnectionManagerWrapper {
+    /// The main connection used for regular Redis commands.
+    pub connection: Arc<TokioMutex<ConnectionManager>>,
+    /// A separate connection dedicated to pub/sub operations.
+    pub pub_sub: Arc<TokioMutex<ConnectionManager>>,
+    /// Atomic flag indicating the connection state (true = connected).
+    pub state: Arc<AtomicBool>, // true = connected
+}
+
+impl ConnectionManagerWrapper {
+    /// Creates a new ConnectionManagerWrapper with the provided
+    /// Redis client.
+    ///
+    /// This method establishes two separate connections to Redis:
+    /// 1. A main connection for regular Redis commands
+    /// 2. A dedicated connection for pub/sub operations
+    ///
+    /// Both connections are wrapped in Tokio mutexes to ensure thread
+    /// safety. The connection state is initially set to connected (true).
+    ///
+    /// # Parameters
+    ///
+    /// - `client` - A reference to a Redis Client instance used to
+    ///    establish connections
+    ///
+    /// # Returns
+    ///
+    /// - `Result<Self, RedisError>` - A new ConnectionManagerWrapper or a
+    ///   Redis error
+    ///
+    /// # Errors
+    ///
+    /// This method will return an error if it fails to establish either
+    /// connection to Redis.
+    pub async fn new(client: &Client) -> Result<Self, RedisError> {
+        // Create connection manager for normal operations
+        let conn = client.get_connection_manager().await?;
+        let pub_sub_conn = client.get_connection_manager().await?;
+        Ok(ConnectionManagerWrapper {
+            connection: Arc::new(TokioMutex::new(conn)),
+            pub_sub: Arc::new(TokioMutex::new(pub_sub_conn)),
+            state: Arc::new(AtomicBool::new(true)),
+        })
+    }
+
+    /// Checks the health of the Redis connection by sending a PING command.
+    ///
+    /// This method attempts to send a PING command to Redis with a timeout
+    /// of 2 seconds. If the PING succeeds, it returns true. If it fails or
+    /// times out, it updates the connection state to disconnected (false)
+    /// and returns false.
+    ///
+    /// # Returns
+    ///
+    /// - `bool` - true if the connection is healthy, false otherwise
+    pub async fn ping(&self) -> bool {
+        let timeout_future = time::timeout(
+            Duration::from_secs(2),
+            AssertUnwindSafe(async {
+                redis::cmd("PING")
+                    .query_async::<_, String>(&mut *self.connection.lock().await)
+                    .await
+            })
+            .catch_unwind(),
```

Comment:
Not sure if you were able to look into the complexity here given all the other work done on the PR.

It is not a lot - I am wondering if necessary, ie I would expect the timeout to be included in the redis client.

Reply:
That is no longer part of the code.